### ISSUE TITLE
Proposal: Require an empty line between "switch-case" statements (newline-between-switch-case)

### ISSUE THREAD

**Issue by monolithed** [eslint/eslint#6619]
## Proposal:

Require an empty line between `switch-break` statements (newline-between-switch-break) #6619
 (`newline-between-switch-break`)
### Options

This rule has an object option:
- `"spaceAfterBreak": true` (default).
- `"spaceBeforeBreak": false`
### Relative
- The curlies should be covered by `padded-blocks`.

``` js
switch (value) {

     case 100000000000000:
          break;

    default:
          break;

}
```
- Comments should be covered by `lines-around-comment`:

For consistency reasons comments are ignored and do not count as empty lines.

``` js
switch (value) {
     case 100000000000000:
          break;

    // comment
    default:
          break;
}
```
### Examples

`/*eslint newline-between-switch-break: "error"*/`
or
`/*eslint newline-between-switch-break: [ "error", {"spaceAfterBreak": true }]*/`

Example of correct code for this rule:

``` js
switch (value) {
     case 100000000000000:
          break;

     case 200000000000000:
          break;

     default:
          break;
 }
```

Example of incorrect code for this rule:

``` js
switch (value) {
     case 100000000000000:
          value
          break;
     case 200000000000000:
          break;
     default:
          break;
 }
```

`/*eslint newline-between-switch-break: ["error", { "spaceBeforeBreak": true }]*/`

``` js
switch (value) {
     case 100000000000000:
          value += 100;

          break;
 }
```

Example of incorrect code for this rule:

``` js
switch (value) {
     case 100000000000000:
          value += 100;
          break;
 }
```

**Comment by platinumazure** [comment]
Just checking, are you looking for an ESLint rule that requires a line break between cases?

**Comment by monolithed** [comment]
Yep

**Comment by platinumazure** [comment]
Weird, I thought we had a rule for this, but my search attempts on the ESLint website unearthed nothing.

Please edit your original post to follow the [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules#proposing-a-rule) guidelines, so we have a better idea of what is being requested here. Thanks!

**Comment by monolithed** [comment]
@platinumazure, done

**Comment by platinumazure** [comment]
@monolithed Thanks, I appreciate the examples.

What about intentional fallthrough? Should those have line breaks in between as well (or perhaps, should that be configurable)?

**Comment by monolithed** [comment]
@platinumazure, done

**Comment by platinumazure** [comment]
Awesome. FYI, I think the `spaceAfterFirstBrace` and `spaceBeforeLastBrace` should be covered by the `padded-blocks` rule.

**Comment by btmills** [comment]
1. I agree with @platinumazure that spacing between the curlies should be covered by `padded-blocks`.
1. It sounds like the goal here is to control the spacing between the `case`s (and the `default`).
    1. In that sense, configuring the newline _before_ a `break` seems to be a separate concern, especially since `break`s can occur elsewhere outside of a `switch`. We have `newline-before-return`. Would it make sense to have a similar rule for `break` and `continue`?
    1. That would reduce the configuration for this rule to just "do I want an empty line between cases that fall through?" and "do I want an empty line between cases that don't fall through?"
    1. If that's all this rule covers, perhaps call it `newline-between-switch-case` instead?

**Comment by monolithed** [comment]
> I think the spaceAfterFirstBrace and spaceBeforeLastBrace should be covered by the padded-blocks rule.

Thanks, fixed

> 1. We have newline-before-return. Would it make sense to have a similar rule for `break` and `continue`?

`spaceAfterBreak` and `spaceBeforeBreak` (reasonable):

```js
switch (value) {
     case 100000000000000:
          break; 

     case 200000000000000:
          value += 1;

          break; 

     default:
          break; 
 }
```

`spaceAfterBreak` and `spaceBeforeBreak` (needlessly):

```js
for (let index of [, ]) {
    if (index++) {
        break;

    } 
}

for (let index of [, ]) {
    if (index++)
        break;

    continue;
}

for (let i of [, ]) {

    break;

}

for (let i of [, ]) 
    break;

for (let i of [, ]); break;
```

This cases are covered by `padded-blocks` and `brace-style`

> That would reduce the configuration for this rule to just "do I want an empty line between cases that fall through?" and "do I want an empty line between cases that don't fall through?"

Maybe we should add the rules for `padded-blocks`? 

> If that's all this rule covers, perhaps call it newline-between-switch-case instead?

Sounds reasonable to me, thanks ! Fixed.

**Comment by btmills** [comment]
Nice examples! That helps clarify the problem. I'm envisioning the `newline-before-return` equivalent for `break` and `continue` as being almost identical to the `return` version, specifically in that it would exclude any statement that is alone in its block.

What about this?

- `"newline-before-flow-control": ["error", "always/never"]`: require an empty line before `break` or `continue` statements
- `"newline-between-switch-case": ["error", "always/never", { "fallthrough": "never"}]`: require an empty line between the `case`s (and `default`), if present) of a `switch` statement. The object argument for fallthrough would only be applicable if the rule were set to `"always"`.

```js
/*eslint newline-before-flow-control: ["error", "always"]*/
/*eslint newline-between-switch-case: ["error", "always", { "fallthrough": "never" }]*/
switch (foo) {
    case a: // newline before (or lack thereof) is controlled by padded-blocks
        doSomething();

        break; // newline before because `break` follows another statement

    case b: // newline before because newline-between-switch-case "always"
        break; // no newline before because `break` is alone in the `case`

    case c: // newline before because newline-between-switch-case "always"
    case d: // no newline before because newline-between-switch-case fallthrough "never"
        doAnotherThing();

        break; // newline before because `break` follows another statement

    default: // newline before because newline-between-switch-case "always"
        throw new Error("This should never happen");
}

for (let i = 0; i < arr.length; i++) {
    if (goodEnough(arr[i])) {
        break; // no newline before because `break` is alone in the block
    }
}

// Your examples with these options:
switch (value) {
     case 100000000000000:
          break; 

     case 200000000000000:
          value += 1;

          break; 

     default:
          break; 
}

for (let index of [, ]) {
    if (index++) {
        break;
    } 
}

for (let index of [, ]) {
    if (index++)
        break;

    continue;
}

for (let i of [, ]) {
    break;
}

for (let i of [, ]) 
    break;

// You had `for (let i of [, ]); break;`, meaning `break` is not in the loop. Was that intentional? I'll do both.
for (let i of [, ]);

break; 

for (let i of [, ]) break;
```

Is that code formatted more like what you want?

**Comment by monolithed** [comment]
@btmills, Sounds reasonable. I like your suggestion.

**Comment by lukeapage** [comment]
I wanted this rule (newline-between-switch-case), but didn't want to wait for this to be accepted, so have implemented in a plugin - https://github.com/lukeapage/eslint-plugin-switch-case
Please feel free to use the code as the basis for putting this rule in core (might need a bit more tidying/documentation/code formatting)

**Comment by mysticatea** [comment]
:+1: I like `newline-between-switch-case` idea. This is similar to #5949.

**Comment by alberto** [comment]
@mysticatea @btmills are you championing this? It doesn't seem critical for me, to be in core.

**Comment by btmills** [comment]
As long as we don't have a conflict here with padded-blocks, I think @lukeapage's plugin has this covered and we can close this! Any objections?

**Comment by monolithed** [comment]
> It doesn't seem so critical for me to be in core.

Why do you thinks so? switch-case-specific rules are a basic functionality and it's strange to see these rules in any third party plugins.

**Comment by cfmano** [comment]
Is there any sign of moving this propsal any further? I find this rules as really usefull ;-)

**Comment by alberto** [comment]
@monolithed from http://eslint.org/docs/developer-guide/contributing/new-rules

> Note: As of 2016, we accept only rules that are deemed extremely important for inclusion. We prefer that new rules be implemented in plugins.

**Comment by monolithed** [comment]
@alberto, could you explain what does mean extremely important rules? And why the switch statement is not one of them...

**Comment by ilyavolodin** [comment]
@monolithed Basically, our bar for entry is "is ESLint incomplete without this rule?". As in, would we miss a bug without this rule?

**Comment by kaicataldo** [comment]
I'm closing this issue because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
